File: tuple.c
#include "tuple.h"

#include "floats.h"

const float POINT_W = 1.0;
const float VECTOR_W = 0.0;
const float COLOR_W = 0.0;

tuple newTuple(float x, float y, float z, float w) {
    tuple t = {x, y, z, w};
    return t;
}

tuple newPoint(float x, float y, float z) { return newTuple(x, y, z, POINT_W); }

tuple newVector(float x, float y, float z) {
    return newTuple(x, y, z, VECTOR_W);
}

tuple newColor(float r, float g, float b) { return newTuple(r, g, b, COLOR_W); }

bool isPoint(tuple t) { return t.w == 1.0; }

bool isVector(tuple t) { return t.w == 0.0; }

bool equalTuple(tuple a, tuple b) {
    return equalFloat(a.x, b.x) && equalFloat(a.y, b.y) &&
           equalFloat(a.z, b.z) && equalFloat(a.w, b.w);
}

tuple addTuple(tuple a, tuple b) {
    a.x += b.x;
    a.y += b.y;
    a.z += b.z;
    a.w += b.w;
    return a;
}

tuple substractTuple(tuple a, tuple b) {
    a.x -= b.x;
    a.y -= b.y;
    a.z -= b.z;
    a.w -= b.w;
    return a;
}

tuple negateTuple(tuple a) {
    a.x *= -1;
    a.y *= -1;
    a.z *= -1;
    a.w *= -1;
    return a;
}

tuple multiplyTuple(tuple a, float s) {
    a.x *= s;
    a.y *= s;
    a.z *= s;
    a.w *= s;
    return a;
}

tuple divideTuple(tuple a, float d) {
    a.x /= d;
    a.y /= d;
    a.z /= d;
    a.w /= d;
    return a;
}

float magnitude(tuple a) { return sqrtf(a.x * a.x + a.y * a.y + a.z * a.z); }

tuple normalize(tuple a) { return divideTuple(a, magnitude(a)); }

float dot(tuple a, tuple b) {
    return a.x * b.x + a.y * b.y + a.z * b.z + a.w * b.w;
}

tuple cross(tuple a, tuple b) {
    tuple x = {a.y * b.z - a.z * b.y, a.z * b.x - a.x * b.z,
               a.x * b.y - a.y * b.x};
    return x;
}

tuple hadamard(tuple a, tuple b) {
    a.x *= b.x;
    a.y *= b.y;
    a.z *= b.z;
    return a;
}

File: go/render/tuple.go
package render

import (
	"math"
)

const EPSILON = 0.00001
const max16bits = 0xffff

type Tuple struct {
	x, y, z, w float64
}

func NewTuple(x, y, z, w float64) Tuple {
	return Tuple{x, y, z, w}
}

func NewPoint(x, y, z float64) Tuple {
	return Tuple{x, y, z, 1}
}

func NewVector(x, y, z float64) Tuple {
	return Tuple{x, y, z, 0}
}

func NewColor(r, g, b float64) Tuple {
	return Tuple{r, g, b, 0}
}

func (a Tuple) Equals(b Tuple) bool {
	return math.Abs(a.x-b.x) < EPSILON &&
		math.Abs(a.y-b.y) < EPSILON &&
		math.Abs(a.z-b.z) < EPSILON &&
		math.Abs(a.w-b.w) < EPSILON
}

func (a Tuple) IsPoint() bool {
	return math.Abs(a.w-1) < EPSILON
}

func (a Tuple) IsVector() bool {
	return math.Abs(a.w) < EPSILON
}

func (a Tuple) X() float64 {
	return a.x
}

func (a Tuple) Y() float64 {
	return a.y
}

func (a Tuple) Z() float64 {
	return a.z
}

func (a Tuple) W() float64 {
	return a.w
}

func (a Tuple) Add(b Tuple) Tuple {
	return Tuple{a.x + b.x, a.y + b.y, a.z + b.z, a.w + b.w}
}

func (a Tuple) Sub(b Tuple) Tuple {
	return Tuple{a.x - b.x, a.y - b.y, a.z - b.z, a.w - b.w}
}

func (a Tuple) Neg() Tuple {
	return Tuple{-a.x, -a.y, -a.z, -a.w}
}

func (a Tuple) Mul(f float64) Tuple {
	return Tuple{a.x * f, a.y * f, a.z * f, a.w * f}
}

func (a Tuple) Div(f float64) Tuple {
	return Tuple{a.x / f, a.y / f, a.z / f, a.w / f}
}

func (a Tuple) Mag() float64 {
	// Taking the magnitude of a point makes no sense, not adding to avoid rounding errors
	return math.Sqrt(a.x*a.x + a.y*a.y + a.z*a.z)
}

func (a Tuple) Norm() Tuple {
	return a.Div(a.Mag())
}

func (a Tuple) Dot(b Tuple) float64 {
	return a.x*b.x + a.y*b.y + a.z*b.z
}

func (a Tuple) Cross(b Tuple) Tuple {
	return Tuple{a.y*b.z - a.z*b.y, a.z*b.x - a.x*b.z,
		a.x*b.y - a.y*b.x, 0}
}

func (a Tuple) Hadamard(b Tuple) Tuple {
	return Tuple{a.x * b.x, a.y * b.y, a.z * b.z, 0}
}

func (a Tuple) RGBA() (r, g, b, alpha uint32) {
	return premultiply(a.x, max16bits), // alpha channel not implemented
		premultiply(a.y, max16bits),
		premultiply(a.z, max16bits),
		uint32(max16bits)

}

func (a Tuple) ZeroW() Tuple {
	return Tuple{a.x, a.y, a.z, 0}
}

func (a Tuple) Reflect(n Tuple) Tuple {
	return a.Sub(n.Mul(2 * a.Dot(n)))
}

func premultiply(c, a float64) uint32 {
	if c < 0 || a < 0 {
		return 0
	}

	ret := uint32(c * a)
	if ret > max16bits {
		return max16bits
	}

	return ret
}
